# Fiancial Statement Analysis Using Python

**Company**: Apple Inc.  
**Period:** 2019–2025  
**Tools:** Python, Pandas, Jupyter Notebook  

This project analyzes a company’s financial statements using Python to evaluate its profitability, liquidity, leverage, and operational efficiency. By calculating key financial ratios and performing trend and year-over-year (YoY) analysis, the project provides insights into the company’s historical performance and overall financial health.

## Project Overview

This project evaluates a company’s financial performance and financial health using historical financial statements and **ratio-based analysis** implemented in Python. The objective is to translate raw accounting data into decision-useful insights from an investment perspective.

The project begins with data loading and cleaning, where raw financial statement data is structured for analysis by setting appropriate indices, standardizing column names, handling missing values, and ensuring consistent data types. Next, in the variables and calculation section, key financial line items such as revenue, operating income, net income, assets, equity, and liabilities are extracted and prepared. These variables form the foundation for ratio computation and trend analysis.

The financial ratio calculations section uses the extracted variables to compute key profitability, liquidity, leverage, and efficiency ratios, with clear formulas and comments linking each metric to its economic interpretation. This is followed by a year-over-year (YoY) analysis that examines trends in margins, returns, liquidity, and capital structure, focusing on consistency, volatility, and directional changes rather than single-year performance.

Finally, the project concludes with an overall conclusion, summarizing the company’s overall financial quality and highlighting strengths, weaknesses, and forward-looking implications.

### Table of Contents
1. [Project Overview](#Project-Overview)  
2. [Data Loading & Cleaning](#Data-Loading-&-Cleaning)  
3. [Variables and Calculation](#Variables-and-Calculation)  
4. [Financial Ratio Calculations](#Financial-Ratio-Calculations)  
5. [Year-over-Year (YoY) Analysis](#Year-over-Year-(YoY)-Analysis) 
6. [Profitability](#profitability) 
7. [Liquidity](#liquidity)
8. [Leverage](#leverage)
9. [Overall-Conclusion](#overall-conclusion)

#### Data Loading & Cleaning

In [1]:
# Imports financial statements, standardizes columns, handles missing values, and prepares data for analysis.

import pandas as pd

df_balancesheet = pd.read_csv("C:\\Users\\hpwal\\Downloads\\New folder (2)\\AAPL_BS.csv", encoding = 'utf-8')
df_pl = pd.read_csv("C:\\Users\\hpwal\\Downloads\\New folder (2)\\AAPL_PL.csv", encoding = 'utf-8')

# Setting Index
df_balancesheet = df_balancesheet.set_index('Unnamed: 0')
df_pl = df_pl.set_index('Unnamed: 0')

# Droppping Nulls
df_balancesheet = df_balancesheet.dropna(how = 'all')
df_pl = df_pl.dropna(how = 'all')

# Converitng data to Numeric
df_balancesheet = df_balancesheet.apply(pd.to_numeric, errors = 'coerce')
df_pl = df_pl.apply(pd.to_numeric, errors = 'coerce')

# Making both tables index same ()
df_balancesheet.columns = df_balancesheet.columns.str.extract(r'(\d{4})')[0]
df_pl.columns = df_pl.columns.str.extract(r'(\d{4})')[0]

# Cleaning index
df_balancesheet.index = df_balancesheet.index.str.strip().str.lower().str.replace(" ",'_')
df_pl.index = df_pl.index.str.strip().str.lower().str.replace(" ","_")
 

#### Varaibles and calculation 

In [2]:
# Extracts key financial items and sets up variables for ratio and trend calculations.

current_assets = df_balancesheet.loc['total_current_assets']
current_liabilities = df_balancesheet.loc['total_current_liabilities']

revenue = df_pl.loc['revenue']
net_income = df_pl.loc['net_income']

total_assets = df_balancesheet.loc['total_assets']
total_equity = df_balancesheet.loc['total_equity']

inventories = df_balancesheet.loc['inventories']


#Calculations (Average)
avg_totalassets = (total_assets + total_assets.shift(1))/2
avg_inventories = (inventories + inventories.shift(1))/2
avg_totalequity = (total_equity + total_equity.shift(1))/2


#Capital Employed
CE = total_assets - current_liabilities 


#### Financial Ratio Calculations

In [3]:
# Current ratio  
# Shows the company’s ability to pay short-term liabilities using current assets.
current_ratio = current_assets / current_liabilities


# Quick ratio 
# Measures ability to cover short-term obligations without relying on inventory
quick = current_assets - df_balancesheet.loc['inventories'] 
quick_ratio = quick / current_liabilities


# Gross margin
# Measures the percentage of revenue retained after covering the cost of goods sold
gross_margin = df_pl.loc['gross_profit'] / revenue


# Operating Profit Ratio 
# Shows the proportion of revenue that remains after operating expenses (SG&A, R&D) are deducted.
Oper_Profit_ratio = df_pl.loc['operating_income_(loss)'] / revenue


# Net Profit Ratio 
# Represents the percentage of revenue converted into net profit.
net_inc_margin = net_income/ revenue


# ROA  
# Measures how efficiently the company uses its assets to generate net profit.
roa = net_income / avg_totalassets


# ROE
# Measures the return generated for shareholders from their invested equity.
roe = net_income /avg_totalequity


# Equity ratio
# Measures the return generated for shareholders from their invested equity.
e_to_q = total_equity / total_assets


# Debt-to-equity
# Shows the proportion of debt used to finance assets relative to equity.
d_to_e = df_balancesheet.loc['long_term_debt'] / total_equity


# Asset turnover
# Measures how efficiently the company uses assets to generate revenue.
asset_turnover = revenue/avg_totalassets


# Receivables turnover 
# Shows how efficiently the company collects its receivables from customers.
Receivables_t = revenue / df_balancesheet.loc['accounts_receivable,_net']


# ROCE ((Return on Capital Emoloyed))
# Measures profitability relative to total capital employed (debt + equity).
roce = df_pl.loc['operating_income_(loss)'] / CE #capital Employed



#### Creating Dataframe for all ratios

In [4]:
fin_health_df = pd.DataFrame({
    'Current ratio': current_ratio,
    'Quick ratio': quick_ratio,
    'Gross Margin': gross_margin,
    'Oper. profit': Oper_Profit_ratio,
    'Net Margin': net_inc_margin,
    'ROA': roa,
    'ROE': roe,
    'Equity ratio': e_to_q,
    'Debt-to-equity': d_to_e,
    'Asset_turnonver': asset_turnover,
    'receivalbes': Receivables_t,
    'ROCE': roce
})

fin_health_df

Current ratio  Quick ratio  Gross Margin  Oper. profit  Net Margin  \
0                                                                          
2019       0.824296     0.803508      0.378178      0.245720    0.212381   
2020       0.704272     0.684371      0.382332      0.241473    0.209136   
2021       1.074553     1.022115      0.417794      0.297824    0.258818   
2022       0.879356     0.847235      0.433096      0.302887    0.253096   
2023       1.059637     1.012909      0.441311      0.298214    0.253062   
2024       0.924504     0.880475      0.462063      0.315102    0.239713   
2025       0.965266     0.927962      0.469052      0.319708    0.269151   

           ROA       ROE  Equity ratio  Debt-to-equity  Asset_turnonver  \
0                                                                         
2019       NaN       NaN      0.267308             NaN              NaN   
2020  0.173341  0.736856      0.201733             NaN         0.828845   
2021  0.280579  1.474433      0.179743        1.729371         1.084079   
2022  0.283629  1.754593      0.143646        1.952933         1.120637   
2023  0.275031  1.719495      0.176259        1.691227         1.086812   
2024  0.261262  1.574125      0.156036        1.697313         1.089897   
2025  0.309325  1.714224      0.205247        1.229816         1.149265   

      receivalbes      ROCE  
0                            
2019    11.348425  0.453433  
2020    17.029467  0.553188  
2021    13.921037  0.483099  
2022    13.991201  0.600871  
2023    12.989189  0.526497  
2024    11.704130  0.617624  
2025    10.462353  0.645999

#### Year-over-Year (YoY) Analysis

In [5]:
#Profitability
gross_margin_yoy = gross_margin.pct_change()*100

oper_margin_yoy = Oper_Profit_ratio.pct_change()*100

net_margin_yoy = net_inc_margin.pct_change()*100

roa_yoy = roa.pct_change()*100

roe_yoy = roe.pct_change()*100

#### Profitability

- **Gross Margin (YoY)**: Remained positive throughout 2020–2025, with growth peaking at ~9.27% in 2021 and moderating to ~1.51% by 2025. This indicates **consistent core profitability** and stable cost control in core operations.

- **Operating Margin (YoY)**: Strong performance in 2021 (~23.34% YoY), followed by a dip in 2023 (~-1.54%) and a moderate recovery in 2025 (~1.46%). The trend reflects **temporary operational volatility**, but cost management appears to be improving in the last two years.

- **Net Margin (YoY)**: Peaked in 2021 (~23.76%), experienced declines in 2022 (~-2.21%) and 2024 (~-5.28%), but rebounded strongly in 2025 (~12.28%). This suggests the company **effectively managed non-operating expenses** and tax impacts over the period.

- **Return on Assets (ROA YoY)**: Showed a significant peak in 2021 (~61.87%), decreased in 2023–2024 (~-3% to -5%), and recovered to ~18.40% in 2025. This indicates **efficient use of assets**, with some sensitivity to operational or non-operating fluctuations.

- **Return on Equity (ROE YoY)**: Highest in 2021 (~100%), followed by a decrease in 2023–2024 (~-2% to -8.45%), and partial recovery to ~8.90% in 2025. The fluctuations suggest that **shareholder returns were affected by leverage and net income volatility**, but have stabilized recently.

### Liquidity

In [6]:
df_liquidity = pd.DataFrame({
    'Current ratio': current_ratio,
    'Quick ratio': quick_ratio,
    'Receivables': Receivables_t
})
df_liquidity

Current ratio  Quick ratio  Receivables
0                                            
2019       0.824296     0.803508    11.348425
2020       0.704272     0.684371    17.029467
2021       1.074553     1.022115    13.921037
2022       0.879356     0.847235    13.991201
2023       1.059637     1.012909    12.989189
2024       0.924504     0.880475    11.704130
2025       0.965266     0.927962    10.462353

**Liquidity Assessment:**
The company’s liquidity position remains tight but controlled throughout the period. The current ratio stays below 1 in most years, ranging from 0.70 (2020) to a **peak of 1.07** (2021), before stabilizing around 0.92–0.97 between 2024–2025. A similar pattern is observed in the quick ratio, **Receivables turnover** remains relatively strong but shows a **gradual decline after 2020**, indicating slower cash collection over time, which slightly pressures liquidity in recent years.

### Leverage

In [7]:
df_leverage = pd.DataFrame({
    'Equity ratio': e_to_q,
    'Debt-to-equity ratio': d_to_e
})

df_leverage.round(2)

Equity ratio  Debt-to-equity ratio
0                                       
2019          0.27                   NaN
2020          0.20                   NaN
2021          0.18                  1.73
2022          0.14                  1.95
2023          0.18                  1.69
2024          0.16                  1.70
2025          0.21                  1.23

**Company Leverage**
The company exhibits high financial leverage, supported by a relatively low equity base. **The equity ratio declines from 0.27 (2019) to a low of 0.14 (2022)**, indicating increased dependence on debt financing during this period. Although equity strength improves modestly afterward, the ratio remains subdued at 0.16 (2024) before recovering to 0.21 (2025).

The company exhibits high financial leverage, supported by a relatively low equity base. The equity ratio declines from 0.27 (2019) to **a low of 0.14 (2022), indicating increased dependence** on debt financing during this period. Although equity strength improves modestly afterward, the ratio remains subdued at 0.16 (2024) before recovering to 0.21 (2025).

#### OVERALL CONCLUSION

The company exhibits a fundamentally strong operating profile, supported by resilient profitability and improving balance-sheet discipline. Profitability metrics indicate effective cost control and capital utilization, with recent recovery in returns reinforcing the strength of the core business. Liquidity remains tight but manageable, reflecting efficient working-capital management despite limited short-term buffers. While leverage has historically been elevated, the recent decline in debt reliance and gradual improvement in equity levels signal a strengthening capital structure. Overall, the company is financially stable, operationally efficient, and well-positioned to support sustainable growth and shareholder value.